### PR TITLE
Fix mindtpy gap check bug

### PR DESCRIPTION
**Pull request by ZedongPeng**
## Summary/Motivation:

In MindtPy, the absolute and relative gap calculation is as follows.
```
solve_data.abs_gap = abs(solve_data.primal_bound - solve_data.dual_bound)
solve_data.rel_gap = solve_data.abs_gap / (abs(solve_data.primal_bound) + 1E-10)
```
Calculating the absolute gap using the `abs()` function might not be appropriate. In the following example(objective sense = maximize), the primal bound is greater than the dual bound in the first iteration. MindtPy should terminate here. However, since `abs()` function is used, `rel_gap` equals `3.27%` instead of `-3.27%`.
Therefore, we should check the objective sense and calculate the absolute and relative gap without `abs()`.
```
---------------------------------------------------------------------------------------------
              Mixed-Integer Nonlinear Decomposition Toolbox in Pyomo (MindtPy)               
---------------------------------------------------------------------------------------------
For more information, please visit https://pyomo.readthedocs.io/en/stable/contributed_packages/mindtpy.html
Original model has 480 constraints (0 nonlinear) and 0 disjunctions, with 144 variables, of which 96 are binary, 0 are integer, and 48 are continuous.
Objective is nonlinear. Moving it to constraint set.
rNLP is the initial strategy being used.

 ===============================================================================================
 Iteration | Subproblem Type | Objective Value | Primal Bound |   Dual Bound |   Gap   | Time(s)

         -       Relaxed NLP           0.27937           -inf        0.27937      nan%      2.51
         1              MILP         0.0416634           -inf      0.0416634      nan%      3.40
*        1         Fixed NLP         0.0430716      0.0430716      0.0416634     3.27%      5.38
         2              MILP         0.0168333      0.0430716      0.0168333    60.92%      6.00
         2         Fixed NLP        -0.0458686      0.0430716      0.0168333    60.92%      6.84
         3              MILP         0.0148974      0.0430716      0.0148974    65.41%      7.70
         3         Fixed NLP        -0.0221965      0.0430716      0.0148974    65.41%      9.44
         4              MILP         0.0140234      0.0430716      0.0140234    67.44%     10.33
         4         Fixed NLP        -0.0230704      0.0430716      0.0140234    67.44%     11.20
         5              MILP         0.0139126      0.0430716      0.0139126    67.70%     11.81
         5         Fixed NLP        -0.0231813      0.0430716      0.0139126    67.70%     12.63
         6              MILP         0.0138067      0.0430716      0.0138067    67.94%     13.18
         6         Fixed NLP        -0.0232871      0.0430716      0.0138067    67.94%     14.51
         7              MILP         0.0136959      0.0430716      0.0136959    68.20%     15.23
         7         Fixed NLP         -0.023398      0.0430716      0.0136959    68.20%     16.04
         8              MILP         0.0133259      0.0430716      0.0133259    69.06%     16.88
         8         Fixed NLP        -0.0470632      0.0430716      0.0133259    69.06%     17.76
         9              MILP         0.0129318      0.0430716      0.0129318    69.98%     18.37
         9         Fixed NLP        -0.0111678      0.0430716      0.0129318    69.98%     19.76
        10              MILP          0.012822      0.0430716       0.012822    70.23%     20.37
        10         Fixed NLP        -0.0242719      0.0430716       0.012822    70.23%     21.25
        11              MILP           0.01282      0.0430716        0.01282    70.24%     21.89
        11         Fixed NLP        -0.0113398      0.0430716        0.01282    70.24%     23.25
        12              MILP         0.0127211      0.0430716      0.0127211    70.47%     23.85
        12         Fixed NLP        -0.0243728      0.0430716      0.0127211    70.47%     24.65
        13              MILP         0.0126337      0.0430716      0.0126337    70.67%     25.28
        13         Fixed NLP        -0.0244602      0.0430716      0.0126337    70.67%     26.23
        14              MILP         0.0125044      0.0430716      0.0125044    70.97%     26.83
        14         Fixed NLP        -0.0245895      0.0430716      0.0125044    70.97%     28.18
        15              MILP          0.012417      0.0430716       0.012417    71.17%     29.03
        15         Fixed NLP        -0.0246769      0.0430716       0.012417    71.17%     29.99
Algorithm is not making enough progress. Exiting iteration loop.
Final bound values: Primal Bound: 0.043071553722709155  Dual Bound: 0.012417005261587122
 ===============================================================================================
 Primal integral          :    0.0000 
 Dual integral            :    0.7267 
 Primal-dual gap integral :    0.7267 
```

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Review by emma58 — CHANGES_REQUESTED** [review]
This looks good @ZedongPeng, but could you please add a test? Perhaps just checking that the example model you posted in the description does indeed converge in one iteration.

**Comment by ZedongPeng** [comment]
Hi @emma58. I added a unit test for the `update_gap` change. The example posted above is not a toy model and is provided by a user of MindtPy. Therefore, adding it to the MindtPy test might not be appropriate.

File: pyomo/contrib/mindtpy/tests/unit_test.py [hunk 1]
```diff
@@ -315,6 +315,13 @@ def test_handle_termination_condition(self):
             self.assertIs(
                 solve_data.results.solver.termination_condition, tc.feasible)
 
+            solve_data.primal_bound = 0.5
+            solve_data.dual_bound = 1
+            update_gap(solve_data)
+            self.assertEqual(solve_data.abs_gap, -0.5)
+            self.assertIs(algorithm_should_terminate(
+                solve_data, config, check_cycling=False), True)
```
Comment (emma58) [inline pyomo/contrib/mindtpy/tests/unit_test.py:323]: You can use `self.assertTrue` here.

**Review by emma58 — APPROVED** [review]
Looks good, thank you for the test!

**Review by jsiirola — APPROVED** [review]

**Pull request merged**